src/spatial_data_format.rs

```diff
@@ -0,0 +1,515 @@
+                Ok(f32::from_be_bytes(four_bytes))
+            })
+            .collect::<Result<Vec<f32>, GrapeFileError>>()?;
+
+        Ok(GrapeFile { header, samples })
+    }
+
+    /// Extract the streams from a [GrapeFile], also returns the sample rate
+    /// because the streams can be encoded at any sample rate.
+    pub fn streams_native_sample_rate(&self) -> (u64, Vec<(GrapeTag, Vec<f32>)>) {
+        let sample_vecs = self.get_raw_streams();
+
+        let res_vecs = Self::attach_tags(&self.header.tags, sample_vecs);
+
+        (self.header.sample_rate, res_vecs)
+    }
+
+    /// Extracts the streams from a [GrapeFile], interpolating or quantizing
+    /// the streams to produce datapoints at the requested sample rate.
+    pub fn streams_with_sample_rate(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        match sample_rate.cmp(&self.header.sample_rate) {
+            Ordering::Equal => self.streams_native_sample_rate().1,
+            Ordering::Less => self.streams_quantized(sample_rate),
+            Ordering::Greater => self.streams_interpolated(sample_rate),
+        }
+    }
+
+    /// Take a slice of [GrapeTag]s and sample vectors and zip them.
+    fn attach_tags(tags: &[GrapeTag], samples: Vec<Vec<f32>>) -> Vec<(GrapeTag, Vec<f32>)> {
+        assert_eq!(tags.len(), samples.len());
+        tags.iter().cloned().zip(samples).collect()
+    }
+
+    /// Returns a cloned, de-interleaved version of the samples in the file.
+    fn get_raw_streams(&self) -> Vec<Vec<f32>> {
+        let n_streams = self.header.n_streams as usize;
+        if n_streams == 0 {
+            return Vec::new();
+        }
+        let stream_len = self.samples.len() / n_streams;
+        let mut sample_vecs = vec![Vec::with_capacity(stream_len); n_streams];
+
+        for (i, &e) in self.samples.iter().enumerate() {
+            let stream_idx = i % n_streams;
+            sample_vecs[stream_idx].push(e);
+        }
+        sample_vecs
+    }
+
+    /// Extracts the streams from the file, and interpolates data points to
+    /// produce data points at the requrested sample_rate.
+    fn streams_interpolated(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        debug_assert!(sample_rate > self.header.sample_rate);
+        let samples_per_pt = sample_rate as usize / self.header.sample_rate as usize;
+        let raw_streams = self.get_raw_streams();
+        let interpolated_streams = raw_streams
+            .into_iter()
+            .map(|v| {
+                v.windows(2)
+                    .flat_map(|w| {
+                        let step = (w[1] - w[0]) / samples_per_pt as f32;
+                        (0..samples_per_pt)
+                            .map(|i| w[0] + i as f32 * step)
+                            .collect::<Vec<_>>()
+                    })
+                    .collect()
+            })
+            .collect();
+        Self::attach_tags(&self.header.tags, interpolated_streams)
+    }
```

Comment:
So if I'm reading this correctly, the requested sample rate must be a multiple of the actual sample rate? Or at least will only yield the requested sample rate if it's a multiple. If that's the case, may be worth mentioning in the function contract

Reply:
Argh I think you are right....may need to rethink how these work because I'd like it to be more flexible than that.

For now I'll document that limitation, we should discuss a better algorithm.